File: android/android_inmobi/src/main/java/com/inmobi/monetization/api/net/JSONPayloadCreator.java
package main.java.com.inmobi.monetization.api.net;

import main.java.com.inmobi.monetization.api.request.ad.Data;
import main.java.com.inmobi.monetization.api.request.ad.Device;
import main.java.com.inmobi.monetization.api.request.ad.Geo;
import main.java.com.inmobi.monetization.api.request.ad.Impression;
import main.java.com.inmobi.monetization.api.request.ad.Property;
import main.java.com.inmobi.monetization.api.request.ad.Request;
import main.java.com.inmobi.monetization.api.request.ad.Slot;
import main.java.com.inmobi.monetization.api.request.ad.User;
import main.java.com.inmobi.monetization.api.request.ad.UserSegment;
import main.java.com.inmobi.monetization.api.request.enums.AdRequest;
import main.java.com.inmobi.monetization.api.request.enums.Gender;
import main.java.com.inmobi.monetization.api.utils.InternalUtil;
import main.java.com.inmobi.monetization.api.utils.LogLevel;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;



public class JSONPayloadCreator {

	/**
	 * This method constructs a JSONObject, depending on the request parameters passed.
	 * @note The request object needs to have valid arguments, depending on the AdRequestType
	 * You should check if the request.isValid() returns true, as this method internally
	 * validates the request & proceeds if & only the request object is valid for the AdRequestType.
	 * @param request The request object, for which a JSON would be constructed
	 * @param type The request type, one of Banner, Interstitial or Native ads
	 * @return JSONObject which is used as POST Body in the InMobi API 2.0 Ad Request. 
	 * TODO User demography as part of JSON API request.
	 */
	public static JSONObject generateInMobiAdRequestPayload(Request request) {

		if (request == null) {
			InternalUtil.Log("Request object cannot be null",LogLevel.ERROR);
			return null;
		}
		if(!request.isValid()) {
			InternalUtil.Log("Please provide valid parameters in the request object",LogLevel.ERROR);
			return null;
		}
		JSONObject mainObject = new JSONObject();
		// request format
		try {
			if (request.getRequestType() != AdRequest.NATIVE) {
				
				mainObject.put("responseformat", "axml");
			} else {
				mainObject.put("responseformat", "native");
			}

			// site format

			JSONObject propertyObject = getPropertyJSON(request.getProperty());
			if (propertyObject != null) {
				
				mainObject.put("site", propertyObject);
			}

			JSONArray impArray = getImpressionJSON(request.getImpression(), request.getRequestType());
			if(impArray != null) {
				mainObject.put("imp", impArray);
			}
			
			JSONObject userObject = getUserJSON(request.getUser());
			if(userObject != null) {
				mainObject.put("user", userObject);
			}
			
			JSONObject deviceObject = getDeviceJSON(request.getDevice());
			if(deviceObject != null) {
				mainObject.put("device", deviceObject);
			}
		} catch(JSONException e) {}

		return mainObject;

	}

	private static JSONObject getPropertyJSON(Property property) throws JSONException {
		JSONObject siteObject = null;
		if (property != null) {
			siteObject = new JSONObject();
			String propertyId = property.getPropertyId();
			if (propertyId != null) {
				siteObject.put("id", propertyId);
			}

		}
		return siteObject;
	}

	private static JSONArray getImpressionJSON(Impression imp,
			AdRequest type) throws JSONException {
		JSONArray impArray = null;
		if (imp != null) {
			JSONObject impressionObject = new JSONObject();

			Slot banner = imp.getSlot();
			if (type != AdRequest.NATIVE) {
				JSONObject bannerObject = new JSONObject();

				String pos = banner.getPosition();
				if (pos != null) {
					bannerObject.put("pos", pos);
				}
				bannerObject.put("adsize", banner.getAdSize());
				impressionObject.put("banner", bannerObject);
			}
			// displaymanager/ver values have a default value, ads too has a
			// default value.
			impressionObject.put("ads", imp.getNoOfAds());
			impressionObject.put("displaymanager",
					imp.getDisplayManager());
			impressionObject.put("displaymanagerver",
					imp.getDisplayManagerVersion());

			// adtype=int, for interstitial ads.
			if (type == AdRequest.INTERSTITIAL) {
				impressionObject.put("adtype", "int");
			}
			// impression object is an array
			impArray = new JSONArray();
			impArray.put(impressionObject);

		}
		return impArray;
	}
	
	private static JSONObject getDeviceJSON(Device device) throws JSONException {
		JSONObject deviceObject = null;
		if(device != null) {
			deviceObject = new JSONObject();
			if (device != null) {
				String ip = device.getCarrierIP();
				if (ip != null) {
					deviceObject.put("ip", ip);
				}
				String ua = device.getUserAgent();
				if (ua != null) {
					deviceObject.put("ua", ua);
				}
				//for Android sites..
				String gpid = device.getGpId();
				if (gpid != null) {
					deviceObject.put("gpid", gpid);
				}
			}

			Geo geo = device.getGeo();
			if (geo != null && geo.isValid()) {
				JSONObject geoObject = new JSONObject();
				geoObject.put("lat", geo.getLatitude());
				geoObject.put("lon", geo.getLongitude());
				geoObject.put("accu", geo.getAccuracy());
				deviceObject.put("geo", geoObject);
			}
		}
		return deviceObject;
	}
	
	private static JSONObject getUserJSON(User user) throws JSONException{
		JSONObject userObject = null;
		if(user != null) {
			userObject = new JSONObject();
			if (user.getYearOfBirth() > 0) {
				userObject.put("yob", user.getYearOfBirth());
			}
			Gender gender = user.getGender();
			if (gender != Gender.UNKNOWN) {
				if (gender == Gender.MALE) {
					userObject.put("gender", "M");
				} else {
					userObject.put("gender", "F");
				}
			}
			JSONObject dataObject = new JSONObject();
			Data data = user.getData();
			if(data != null) {
				String ID = data.getID();
				if (ID != null) {
					dataObject.put("id", ID);
				}
				String name = data.getName();
				if (name != null) {
					dataObject.put("name", name);
				}
				UserSegment segmentObject = data.getUserSegment();
				if (segmentObject != null) {
					if (segmentObject.getUserSegmentArray() != null) {
						// dataObject.put("segment",new JSONObject());
						JSONArray dataArray = new JSONArray();
						dataArray.put(dataObject);
						userObject.put("data", dataArray);
					}
				}
			}
		}
		return userObject;
	}

}

File: php/inmobi/monetization/ads/api/request/JSONPayloadCreator.php
<?php
/////////////////////////////////////////////////////////////////
// Copyright © 2015 InMobi Technologies Pte. Ltd. All rights reserved.

//                           MIT License

//Permission is hereby granted, free of charge, to any person obtaining a copy
//of this software and associated documentation files (the "Software"), to deal
//in the Software without restriction, including without limitation the rights
//to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//copies of the Software, and to permit persons to whom the Software is
//furnished to do so, subject to the following conditions:

//The above copyright notice and this permission notice shall be included in
//all copies or substantial portions of the Software.

//THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//THE SOFTWARE.
////////////////////////////////////////////////////////////////////////////// 
require_once(dirname(__FILE__)."/Enums.php");
require_once(dirname(__FILE__)."/Request.php");
/**
* This class is used by Banner, Interstitial or Native moentization classes to generate
* a JSON payload used for making the POST body for InMobi ad request.
*/
class JSONPayloadCreator {

	/**
	 * This method constructs a json object, depending on the request parameters
	 * passed.
	 * 
	 * @note The request object needs to have valid arguments, depending on the
	 *       AdRequestType Please check if the request.isValid() returns
	 *       true, as this method internally validates the request & proceeds if
	 *       & only the request object is valid for the AdRequestType.
	 * @param $request
	 *            The request object, for which a JSON would be constructed
	 * @return Json Object which is used as POST Body in the InMobi API 2.0 Ad
	 *         Request. TODO User demography as part of JSON API request.
	 */
	public static function generateInMobiAdRequestPayload($request) {
		if($request == null || gettype($request) != "object") return null;
		if($request->isValid() != true) return null;

		$mainObjectArray = array();

		if($request->requestType == AdRequest::NATIVE) {
			$mainObjectArray['responseformat'] = "native";
		} else {
			$mainObjectArray['responseformat'] = "axml";
		}
		
		$propertyJson = JSONPayloadCreator::getPropertyJson($request->property);
		if($propertyJson != null) {
			$mainObjectArray['site'] = $propertyJson;
		}

		$impJson = JSONPayloadCreator::getImpressionJson($request->impression,$request->requestType);
		if($impJson != null) {
			$mainObjectArray['imp'][] = $impJson;
		}
		$deviceJson = JSONPayloadCreator::getDeviceJson($request->device);
		if($deviceJson != null) {
			$mainObjectArray['device'] = $deviceJson;
		}

		$userJson = JSONPayloadCreator::getUserJson($request->user);
		if($userJson != null && empty($userJson) == false) {
			$mainObjectArray['user'] = $userJson;
		}

		return json_encode($mainObjectArray);
	}

	private static function getPropertyJson($property) {
		$propArray = array();
		$propArray['id'] = $property->propertyId;
		return $propArray;
	}

	private static function getImpressionJson($impression,$requestType) {

		$impArray = array();
		$impArray['ads'] = $impression->noOfAds;
		$impArray['displaymanager'] = $impression->displayManager;
		$impArray['displaymanagerver'] = $impression->displayManagerVersion;
		if($requestType == AdRequest::INTERSTITIAL) {
			$impArray['adtype'] = "int";
		}

		if($requestType != AdRequest::NATIVE) {
			$bannerArray = array();
			$slot = $impression->slot;
			$bannerArray['adsize'] = $slot->adSize;
			if(StubValidator::isStringValid($slot->position)) {
				$bannerArray['pos'] = $slot->position;
			}
			$impArray['banner'] = $bannerArray;
		}

		return $impArray;
	}
	private static function getDeviceJson($device) {
		$deviceArray = array();
		$deviceArray['ip'] = $device->carrierIP;
		$deviceArray['ua'] = $device->userAgent;

		//anddroid identifiers
		$gpid = $device->gpId;
		if(StubValidator::isStringValid($gpid)) {
			$deviceArray['gpid'] = $gpid;
		}
		//aid here must be a SHA-1 hashed value of the original ANDROID_ID, or otherwise should be 
		//passed as md5 value..
		// if you don't know this, do not pass this identifier, but rather pass google advertising identifier.
		$aid = $device->androidId;
		if(StubValidator::isStringValid($aid)) {
			$deviceArray['o1'] = $aid;
		}

		//iOS identifiers
		$ida = $device->idfa;
		if(StubValidator::isStringValid($ida)) {
			$deviceArray['ida'] = $ida;
		}
		$adt = $device->adTrackingDisabled;
		if($adt == true) {
			$deviceArray['adt'] = 1;
		} else {
			$deviceArray['adt'] = 0;
		}
		//geo object
		$geo = $device->geo;
		if($geo != null && $geo->isValid() == true) {
			$geoArray = array();
			$geoArray['lat'] = $geo->lat;
			$geoArray['lon'] = $geo->lon;
			$geoArray['accu'] = $geo->accuracy;
			$deviceArray['geo'] = $geoArray;
		}

		return $deviceArray;
	}

	private static function getUserJson($user) {
		//TODO Data object implementation
		$userArray = array();
		if($user == null || gettype($user) != "object") return $userArray;
		
		$yob = $user->yearOfBirth;
		if(StubValidator::isIntegerValid($yob) && $yob > 0) {
			$userArray['yob'] = $yob;
		}
		$gender = $user->gender;
		if($gender != null) {
			if($gender == Gender::MALE) {
				$userArray['gender'] = 'M';
			} else if($gender == Gender::FEMALE) {
				$userArray['gender'] = 'F';
			}
		}
		return $userArray;
	}

}

?>
